node/src/utils/pidfile.rs

```diff
@@ -0,0 +1,255 @@
+#[derive(Debug)]
+pub enum PidfileOutcome {
+    /// Another instance of the node is likely running, or an attempt was made to reuse a pidfile.
+    ///
+    /// **Recommendation**: Exit to avoid resource conflicts.
+    AnotherNodeRunning(PidfileError),
+    /// The node crashed previously and could potentially have been corrupted.
+    ///
+    /// **Recommendation**: Run an integrity check, then potentially continue with intialization.
+    ///                     **Store the `Pidfile`**.
+    Crashed(Pidfile),
+    /// Clean start, pidfile lock acquired.
+    ///
+    /// **Recommendation**: Continue with intialization, but **store the `Pidfile`**.
+    Clean(Pidfile),
+    /// There was an error managing the pidfile, not sure if we have crashed or not.
+    ///
+    /// **Recommendation**: Exit, as it will not be possible to determine a crash at the next
+    /// start.
+    PidfileError(PidfileError),
+}
+
+impl Pidfile {
+    /// Acquire a `Pidfile` and give an actionable outcome.
+    ///
+    /// **Important**: This function should be called **before** opening whatever resources it is
+    /// protecting.
+    pub fn acquire<P: AsRef<Path>>(path: P) -> PidfileOutcome {
+        match Pidfile::new(path) {
+            Ok(pidfile) => {
+                if pidfile.unclean_shutdown() {
+                    PidfileOutcome::Crashed(pidfile)
+                } else {
+                    PidfileOutcome::Clean(pidfile)
+                }
+            }
+            Err(err @ PidfileError::LockFailed(_)) => PidfileOutcome::AnotherNodeRunning(err),
+            Err(err) => PidfileOutcome::PidfileError(err),
+        }
+    }
+
+    /// Creates a new pidfile.
+    ///
+    /// The error-behavior of this function is important and can be used to distinguish between
+    /// different conditions according to the table below. If the `Pidfile` is instantiated before
+    /// the resource it is supposed to protect, the following actions are recommended:
+    fn new<P: AsRef<Path>>(path: P) -> Result<Pidfile, PidfileError> {
+        // First we try to open the pidfile, without disturbing it.
+        let mut pidfile = fs::OpenOptions::new()
+            .truncate(false)
+            .create(true)
+            .read(true)
+            .write(true)
+            .open(path.as_ref())
+            .map_err(PidfileError::CouldNotOpen)?;
+
+        // Now try to acquire an exclusive lock. This will fail if another process or another
+        // instance of `Pidfile` is holding a lock onto the same pidfile.
+        pidfile
+            .try_lock_exclusive()
+            .map_err(PidfileError::LockFailed)?;
+
+        // At this point, we're the exclusive users of the file and can read its contents.
+        let mut raw_contents = String::new();
+        pidfile
+            .read_to_string(&mut raw_contents)
+            .map_err(PidfileError::ReadFailed)?;
+
+        // Note: We cannot distinguish an empty file from a non-existing file, unfortunately.
```

Comment:
Wouldn't `Path::is_file()` used at the start of this function achieve that?

Reply:
We have to open it to lock it, while creating it if missing. It's certainly to possible to do all this, but not in a race-free manner. The only alternative I can see is to create two files, a pidfile and a lockfile, but that's a whole different can of worms.

Reply:
Yes, fair point.  The scope for a race is small, but not zero.  I guess we could check the creation timestamp after opening, but all this seems like overkill.  I'm happy to leave as is.